cirq-core/cirq/ops/state_preparation_gate.py

```diff
@@ -0,0 +1,125 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Quantum gates defined by a matrix."""
+
+from typing import Any, Sequence, Dict, List, Tuple, TYPE_CHECKING
+
+import numpy as np
+
+from cirq import protocols
+from cirq.ops import raw_types
+from cirq.ops.common_channels import ResetChannel
+from cirq.ops.matrix_gates import MatrixGate
+from cirq._compat import proper_repr
+
+if TYPE_CHECKING:
+    import cirq
+
+
+class PrepareState(raw_types.Gate):
+    """A unitary qubit gate which resets all qubits to the |0> state
+    and then prepares the target state."""
+
+    def __init__(self, target_state: np.ndarray, name: str = "StatePreparation") -> None:
+        """Initializes a matrix gate.
+
+        Args:
+            target_state: The state vector that this gate should prepare.
+            name: the name of the gate
+
+        Raises:
+            ValueError: if the array is not 1D, or does not have 2**n elements for some n.
+        """
+        if len(target_state.shape) != 1:
+            raise ValueError('`target_state` must be a 1d numpy array.')
+
+        n = int(np.round(np.log2(target_state.shape[0] or 1)))
+        if 2 ** n != target_state.shape[0]:
+            raise ValueError(f'Matrix width ({target_state.shape[0]}) is not a power of 2')
+
+        self._state = target_state.astype(np.complex) / np.linalg.norm(target_state)
+        self._num_qubits = n
+        self._name = name
+        self._qid_shape = (2,) * n
+
+    def _json_dict_(self) -> Dict[str, Any]:
+        """Converts the gate object into a serializable dictionary"""
+        return {
+            'cirq_type': self.__class__.__name__,
+            'target_state': self._state.tolist(),
+        }
+
+    def _num_qubits_(self):
+        return self._num_qubits
+
+    @staticmethod
+    def _has_unitary_() -> bool:
+        """Checks and returns if the gate has a unitary representation.
+        It doesn't, since the resetting of the channels is a non-unitary operations,
+        it involves measurement."""
+        return False
+
+    @classmethod
+    def _from_json_dict_(cls, target_state, **_kwargs):
+        """Recreates the gate object from it's serialized form
+
+        Args:
+            target_state: the state to prepare using this gate
+            _kwargs: other keyword arguments, ignored
+        """
+        return cls(target_state=np.array(target_state))
+
+    def _qid_shape_(self) -> Tuple[int, ...]:
+        return self._qid_shape
+
+    def _circuit_diagram_info_(
+        self, _args: 'cirq.CircuitDiagramInfoArgs'
+    ) -> 'cirq.CircuitDiagramInfo':
+        """Returns the information required to draw out the circuit diagram for this gate."""
+        symbols = (
+            [self._name]
```

Comment:
I have mixed feelings about how this obscures the target state, but also recognize that the state is probably too large to include in the diagram. What are your thoughts on this?

Reply:
Had no other ideas, there are 2^n values, writing inline all the complex number will be too messy even for 2 qubit states. Any better suggestions here would be great.

Reply:
There are a few other issues open revolving around state representation, but I don't think any have offered a general solution for this. For now, "StatePreparation" / user-defined names seems reasonable.